### PR TITLE
added namespace support for metrics

### PR DESCRIPTION
**Pull request by rathijeetbhave**
**Issue :**
This pull request adds namespace support for circuit metrics. If there are two services serviceA and serviceB both using prometrics, then plotting say circuit_timeout_duration_seconds{} would give metrics for both serviceA and serviceB. To view metrics of a specific circuit we have to provide the name parameter in the query. Separate graphs need to be created for every circuit in serviceA. This feature will give an option to user to specify a namespace for service metrics. That way circuit_timeout_duration_seconds{} of all circuits in serviceA can be plotted on the same graph with separate legends without having any data from serviceB.

**Approach :**
A new function is created to initialise histograms and counters which takes an optional namespace parameter. This parameter is added in all histograms and counters along with string **"circuit"** as a sub-system. Existing implementations will continue to work using empty string as namespace. With the new change all metrics will be created under given namespace. EG : **{Namespace}_circuit_timeout_duration_seconds_bucket{}**

### CODE REVIEW

File: metrics.go [hunk 1]
```diff
@@ -20,27 +20,29 @@ var (
...
+	subSystem        = "circuit"
 	circuitNameLabel = "name"
 	funcTypeLabel    = "func"
 
 	runCmd      = "run"
 	fallbackCmd = "fallback"
 )
 
-func init() {
+func newMetrics(nameSpace string) {
```
Comment (jiacai2050) [inline metrics.go]: 1. Could we name this function `initMetrics`? It makes more sense to me.
2. the name of arg should be `namespace`

File: prometrics_test.go [hunk 2]
```diff
@@ -100,3 +100,74 @@ func TestCommandFactory(t *testing.T) {
 		time.Sleep(1 * time.Hour)
 	}
 }
+
+
+func TestCommandFactoryWithNameSpace(t *testing.T) {
```
Comment (jiacai2050) [inline prometrics_test.go]: This test can't test namespace. 

I think you don't need to add a new test for this.

**Review by jiacai2050 — CHANGES_REQUESTED** [review]

File: prometrics.go [hunk 3]
```diff
@@ -111,6 +111,29 @@ var (
 // GetFactory returns a global-shared CommandFactory, with metrics registered into r
 func GetFactory(r prometheus.Registerer) *CommandFactory {
 	once.Do(func() {
+		initMetrics("")
```
Comment (jiacai2050) [inline prometrics.go]: Could we just call `GetFactoryWithNameSpace("", r)` inside GetFactory?

**Review by jiacai2050 — CHANGES_REQUESTED** [review]

File: prometrics.go [hunk 4]
```diff
@@ -111,6 +111,29 @@ var (
...
+		)
+		inst = new(CommandFactory)
+	})
+
+	return inst
+}
+
+// GetFactory returns a global-shared CommandFactory, with metrics registered into r
+// Metrics will be created in particular nameSpace
+func GetFactoryWithNameSpace(r prometheus.Registerer, namespace string) *CommandFactory {
```
Comment (jiacai2050) [inline prometrics.go]: GetFactoryWithNamespace is a better name since namespace is a word.

**Review by jiacai2050 — CHANGES_REQUESTED** [review]

File: prometrics.go (hunk 3 above)
Comment (rathijeetbhave) [reply prometrics.go]: +1

**Comment by jiacai2050** [comment]
Thanks. Release version v0.1.0

**Pull request merged**